lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2265,27 +2275,56 @@ static bool diagnoseUnsatisfiedRequirements(ADContext &context,
   SmallVector<Requirement, 2> unsatisfiedRequirements;
   for (auto req : requirements) {
     auto firstType = req.getFirstType();
-    auto secondType = req.getSecondType();
+    Type secondType;
     // Substitute first and second types using the given substitution map,
     // looking up conformances in the current module, if possible.
     if (auto substFirstType =
             firstType.subst(QuerySubstitutionMap{substMap},
                             LookUpConformanceInModule(swiftModule))) {
       firstType = substFirstType;
     }
-    if (auto substSecondType =
-            secondType.subst(QuerySubstitutionMap{substMap},
-                             LookUpConformanceInModule(swiftModule))) {
-      secondType = substSecondType;
+    if (req.getKind() != RequirementKind::Layout) {
+      secondType = req.getSecondType();
+      if (auto substSecondType =
+              secondType.subst(QuerySubstitutionMap{substMap},
+                               LookUpConformanceInModule(swiftModule))) {
+        secondType = substSecondType;
+      }
     }
     switch (req.getKind()) {
+    // Check layout requirements.
```

Comment:
Note: I couldn't think of ways to test unsatisfied layout requirements. This should be easier after [TF-864](https://bugs.swift.org/browse/TF-864), when `@differentiable` attribute supports layout requirements in where clauses.